### PR TITLE
[RFC FS-1004] FSharpResult type with a few tests

### PR DESCRIPTION
**Pull request by wallymathieu**
This is the implementation of [RFC FS-1004 Result type](https://github.com/fsharp/FSharpLangDesign/blob/master/RFCs/FS-1004-result-type.md)

### CODE REVIEW

**Comment by enricosada** [comment]
@wallymathieu can you add `[WIP]` to title?

File: src/fsharp/FSharp.Core/control.fs [hunk 1]
```diff
@@ -601,7 +601,7 @@ namespace Microsoft.FSharp.Control
         let mutable defaultCancellationTokenSource = new CancellationTokenSource()
 
         [<NoEquality; NoComparison>]
-        type Result<'T>  =
+        type AsyncImplResult<'T>  =
```
Comment (forki) [inline src/fsharp/FSharp.Core/control.fs]: Is this a public type?

File: src/fsharp/FSharp.Core.Unittests/FSharp.Core.Unittests.fsproj [hunk 2]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (enricosada) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core.Unittests.fsproj]: to minimize diff revert this line

File: src/fsharp/FSharp.Core.Unittests/FSharp.Core.Unittests.fsproj [hunk 3]
```diff
@@ -84,7 +84,7 @@
     <Compile Include="FSharp.Core\Microsoft.FSharp.Collections\HashIdentityModule.fs" />
     <Compile Include="FSharp.Core\Microsoft.FSharp.Collections\ListModule.fs" />
     <Compile Include="FSharp.Core\Microsoft.FSharp.Collections\ListModule2.fs" />
-    <Compile Include="FSharp.Core\Microsoft.FSharp.Collections\ObsoleteListFunctions.fs" />    
+    <Compile Include="FSharp.Core\Microsoft.FSharp.Collections\ObsoleteListFunctions.fs" />
```
Comment (enricosada) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core.Unittests.fsproj]: to minimize diff revert this line

File: src/fsharp/FSharp.Core.Unittests/FSharp.Core.Unittests.fsproj [hunk 4]
```diff
@@ -99,13 +99,14 @@
     <Compile Include="FSharp.Core\Microsoft.FSharp.Collections\StringModule.fs" />
     <Compile Include="FSharp.Core\PrimTypes.fs" />
     <Compile Include="FSharp.Core\ComparersRegression.fs" />
-    <Compile Include="FSharp.Core\DiscrimantedUnionType.fs" /> 
+    <Compile Include="FSharp.Core\DiscrimantedUnionType.fs" />
```
Comment (enricosada) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core.Unittests.fsproj]: to minimize diff revert this line

File: src/fsharp/FSharp.Core.Unittests/FSharp.Core.Unittests.fsproj [hunk 5]
```diff
@@ -121,6 +122,6 @@
     <CustomCopyLocal Include="FSharp.Core.Unittests.dll.config">
       <TargetFilename>FSharp.Core.Unittests.dll.config</TargetFilename>
     </CustomCopyLocal>
-    </ItemGroup>
+  </ItemGroup>
```
Comment (enricosada) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core.Unittests.fsproj]: to minimize diff revert this line

File: src/fsharp/FSharp.Core.Unittests/FSharp.Core.Unittests.fsproj [hunk 6]
```diff
@@ -121,6 +122,6 @@
...
   <Import Project="$(FSharpSourcesRoot)\FSharpSource.targets" />
-</Project>
+</Project>
```
Comment (enricosada) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core.Unittests.fsproj]: to minimize diff revert this line

**Comment by forki** [comment]
How does this relate to Chessie? Are you going to port the tests over? If we really put this into FSharp.Core then it should at least do what Chessie can. Especially in regards to C# interop this is a really hard topic. I tried so many things to make Chessie work for C# - so please also port the C# samples / tests to show that we have at elast that basic C# support that Chessie has.

**Comment by enricosada** [comment]
@forki i think it's good to have a shared (in whole f# ) `Success/Failure` type.
But I think is wrong to add more functions in FSharp.Core. 
External libraries like Chessie can add additional functions using the `Result` type inside FSharp.Core

**Comment by forki** [comment]
/// <summary>Helper type for error handling without exceptions.</summary>
    [<StructuralEquality; StructuralComparison>]
    [<CompiledName("FSharpResult`2")>]
    type Result<'T1,'T2> = 
    | Success of 'T1 
    | Error of 'T2
 
Is that really the type we want for Results? Chessie is using:

    /// Represents the result of a computation.
    type Result<'TSuccess, 'TMessage> = 
    /// Represents the result of a successful computation.
    | Ok of 'TSuccess * 'TMessage list
    /// Represents the result of a failed computation.
    | Bad of 'TMessage list

**Comment by enricosada** [comment]
I'll open an RFC to discuss it

**Comment by enricosada** [comment]
@wallymathieu can you add "Discuss on fsharp/FSharpLangDesign#49" in the pr description?
meanwhile i'll add the RFC doc

File: src/fsharp/FSharp.Core/control.fs (hunk 1 above)
Comment (wallymathieu) [inline src/fsharp/FSharp.Core/control.fs]: This is a private type. It's not exposed from control.fsi.

**Comment by wallymathieu** [comment]
@enricosada there, I've updated the description and title.

**Comment by radekm** [comment]
```fsharp
/// Represents the result of a computation.
type Result<'TSuccess, 'TMessage> = 
    /// Represents the result of a successful computation.
    | Ok of 'TSuccess * 'TMessage list
    /// Represents the result of a failed computation.
    | Bad of 'TMessage list
```

I don't see why `Bad` should always contain a `list` of values? It seems suboptimal for functions which return only a single bad value. Additionally `list` has slow concatenation so combining lists with messages would be slow.

**Comment by radekm** [comment]
On the other hand

```fsharp
type Result<'T1,'T2> = 
    | Success of 'T1 
    | Error of 'T2
```

is isomorphic to `Choice<'T1, 'T2>`. So I don't see why we should add it.

File: src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Core/ResultTests.fs [hunk 7]
```diff
@@ -0,0 +1,44 @@
+﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+// Various tests for:
+// Microsoft.FSharp.Core.ExtraTopLevelOperators.printf
```
Comment (mexx) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Core/ResultTests.fs]: Please adjust this comment to match the test target.

**Comment by KevinRansom** [comment]
@wallymathieu @forki @dsyme  @otawfik-ms 
Guys,
where are we on this PR?  it hasn't been touched since mid-feb and so looks like it might be dead.  Do we want to add this functionality, shall we find someone to shepherd it through or drop it?

Kevin

File: src/fsharp/FSharp.Core/prim-types.fsi [hunk 8]
```diff
@@ -1772,6 +1772,13 @@ namespace Microsoft.FSharp.Core
     and 'T option = Option<'T>
 
 
+    /// <summary>Helper type for error handling without exceptions.</summary>
+    [<StructuralEquality; StructuralComparison>]
+    [<CompiledName("FSharpResult`2")>]
+    type Result<'T1,'T2> = 
+      | Success of 'T1
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/prim-types.fsi]: Please rename this to ``Ok``

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 9]
```diff
@@ -4735,6 +4735,12 @@ namespace Microsoft.FSharp.Core
     and 'T option = Option<'T> 
 
 
+    [<StructuralEquality; StructuralComparison>]
+    [<CompiledName("FSharpResult`2")>]
+    type Result<'T1,'T2> = 
+        | Success of 'T1
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/prim-types.fs]: Please rename this to ``Ok``

File: src/fsharp/FSharp.Core/prim-types.fsi (hunk 8 above)
Comment (dsyme) [inline src/fsharp/FSharp.Core/prim-types.fsi]: Please rename the type variables to ``T`` and ``TError``

File: src/fsharp/FSharp.Core/prim-types.fsi [hunk 10]
```diff
@@ -1772,6 +1772,13 @@ namespace Microsoft.FSharp.Core
     and 'T option = Option<'T>
 
 
+    /// <summary>Helper type for error handling without exceptions.</summary>
+    [<StructuralEquality; StructuralComparison>]
+    [<CompiledName("FSharpResult`2")>]
+    type Result<'T1,'T2> = 
+      | Success of 'T1 
+      | Error of 'T2
+
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/prim-types.fsi]: Please add ``///`` comments to  each union case

**Comment by dsyme** [comment]
Please make the adjustments mentioned above and then we can accept this,

There is discussion in the RFC about whether we should also add a corresponding module of operations and a ``attempt`` builder into FSharp.Core as well.  But minimally there is real value in accepting the definition of the type to help enable interop between different F# components that want to use this type.

**Comment by wallymathieu** [comment]
Should I do a merge from master and fix the compilation error in Commands.fs?

**Comment by dsyme** [comment]
@wallymathieu Yes please

**Comment by wallymathieu** [comment]
The failure of ```Windows_NT Release_ci_part2 Build``` : 
FSharp.Core.Unittests.FSharp_Core.Microsoft_FSharp_Control.AsyncModule.Async.AwaitWaitHandle does not leak memory
Does not look related to this pull request.

**Comment by dsyme** [comment]
CLose/reopen to re-run CI

**Comment by wallymathieu** [comment]
Should we change the title and remove [WIP] ?

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 11]
```diff
@@ -3470,6 +3470,12 @@ namespace Microsoft.FSharp.Core
     and 'T option = Option<'T> 
 
 
+    [<StructuralEquality; StructuralComparison>]
+    [<CompiledName("FSharpResult`2")>]
+    type Result<'TOk,'TError> =
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/prim-types.fs]: Please change to just "T" instead of "TOk"

File: src/fsharp/FSharp.Core/prim-types.fsi [hunk 12]
```diff
@@ -1770,6 +1770,15 @@ namespace Microsoft.FSharp.Core
     and 'T option = Option<'T>
 
 
+    /// <summary>Helper type for error handling without exceptions.</summary>
+    [<StructuralEquality; StructuralComparison>]
+    [<CompiledName("FSharpResult`2")>]
+    type Result<'TOk,'TError> =
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/prim-types.fsi]: Likewise here

**Comment by dsyme** [comment]
@wallymathieu removed, thanks.  Added one comment above TOk --> T

**Comment by dsyme** [comment]
This is ready to merge.

**Comment by otawfik-ms** [comment]
LGTM :+1:

**Comment by cartermp** [comment]
👍

**Pull request merged**